Cache expensive operations in Post transformer

pfefferle opened the pull request:
## Proposed changes:

This PR adds caching for expensive operations in the Post transformer to improve performance by avoiding redundant computation during object serialization.

**Cached properties:**
* content: Processed ActivityPub content with shortcodes and filters
* summary: Post summary/excerpt
* tags: Post tags including hashtags
* attachment: Post attachments
* mentions: Extracted mentions from content
* in_reply_to: Reply-to URLs from reply blocks

All cached properties use `false` as a sentinel value to indicate "not yet computed", providing a consistent pattern across all properties and correctly handling nullable return values.

**Content length check removed:**
The content length check in `get_type()` has been removed due to inconsistent usage. Previously, the type determination would sometimes use raw content and sometimes use processed content depending on call order, leading to unpredictable behavior. The type is now determined solely based on post format, title support, and post type.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

1. Checkout the branch
2. Run the test suite: `npx wp-env run tests-cli -- /var/www/html/wp-content/plugins/activitypub/vendor/bin/phpunit -c /var/www/html/wp-content/plugins/activitypub/phpunit.xml.dist`
3. Verify all 1443 tests pass
4. Test creating ActivityPub posts to ensure content, tags, and attachments are correctly generated
5. Verify performance improvement when accessing Post transformer properties multiple times
6. Confirm that short posts with titles are now correctly identified as Articles (not Notes based on length)

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Added - for new features
-   [x] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Cache expensive operations in Post transformer to improve performance.

</details>

pfefferle committed:
Cache expensive operations in Post transformer

Add caching for frequently accessed properties in the Post transformer to improve performance by avoiding redundant computation during object serialization.

Cached properties:
- content: Processed ActivityPub content with shortcodes and filters
- summary: Post summary/excerpt
- tags: Post tags including hashtags
- attachment: Post attachments
- mentions: Extracted mentions from content
- in_reply_to: Reply-to URLs from reply blocks

All cached properties use false as sentinel value to indicate "not yet computed", providing consistent pattern across all properties and correctly handling nullable return values.

Test updates:
- Updated test expectation for short content to return Article instead of Note, reflecting the removal of content length check from get_type()

matticbot committed:
Add changelog

pfefferle commented on includes/transformer/class-post.php:
```diff
@@ -381,14 +431,10 @@ protected function get_type() {
-			\strlen( $content ) <= ACTIVITYPUB_NOTE_LENGTH
+			! \post_type_supports( $this->item->post_type, 'title' ) ||
+			! $this->item->post_title
 		) {
```
this is on purpose, because it used different content versions.

pfefferle commented:
/cc @Menrath

obenland commented on includes/transformer/class-post.php:
Should this be assigned to `$this->attachment`?

obenland commented on includes/transformer/class-post.php:
```diff
@@ -381,14 +431,10 @@ protected function get_type() {
-			\strlen( $content ) <= ACTIVITYPUB_NOTE_LENGTH
+			! \post_type_supports( $this->item->post_type, 'title' ) ||
+			! $this->item->post_title
 		) {
```
```suggestion
// Check if the post has no title.
if ( ! \post_type_supports( $this->item->post_type, 'title' ) || ! $this->item->post_title ) {
```

obenland commented on includes/transformer/class-post.php:
```diff
@@ -464,12 +516,20 @@ protected function get_summary() {
 
+		if ( false !== $this->summary ) {
+			return $this->summary;
+		}
```
There's a difference in behavior with `get_content()`, where drafts are calculated on every call, whereas here this conditional prevents that.

Not sure what's more correct but I assume we'd want to be consistent?

obenland commented on includes/transformer/class-post.php:
```diff
@@ -596,14 +667,17 @@ protected function get_in_reply_to() {
 
-		return \array_values( \array_unique( $reply_urls ) );
+		$this->in_reply_to = \array_values( \array_unique( $reply_urls ) );
+		return $this->in_reply_to;
```
```suggestion

		return $this->in_reply_to;
```

obenland commented on includes/transformer/class-post.php:
```diff
@@ -596,14 +667,17 @@ protected function get_in_reply_to() {
 		if ( empty( $reply_urls ) ) {
-			return null;
+			$this->in_reply_to = null;
+			return $this->in_reply_to;
```
```suggestion

			return $this->in_reply_to;
```

obenland commented on includes/transformer/class-post.php:
```diff
@@ -596,14 +667,17 @@ protected function get_in_reply_to() {
 		if ( 1 === count( $reply_urls ) ) {
-			return \current( $reply_urls );
+			$this->in_reply_to = \current( $reply_urls );
+			return $this->in_reply_to;
```
```suggestion

			return $this->in_reply_to;
```

pfefferle committed:
Merge branch 'trunk' into add/cache-expensive-operations

pfefferle committed:
Refactor post content and reply handling logic

Simplifies conditional checks for post title support and refactors the get_content method to cache content earlier. Also improves in_reply_to assignment by adding explicit returns after setting values.

pfefferle committed:
Fix attachment reset when max_media is zero

Ensures that the attachment property is explicitly set to an empty array when no media attachments are allowed, improving consistency in the return value.

pfefferle committed:
Fix transformer instance caching in test

Creates a new Post transformer instance in the test to avoid using a cached attachment result, ensuring accurate test behavior.

pfefferle committed:
Merge branch 'trunk' into add/cache-expensive-operations
